Make SVGs and CSS dynamically recolourable

ara4n opened the pull request:
I'm not proud of this.

Turns any colour-themed SVGs from img tags into object tags to make them dynamically editable from JS, and implements a simple /tint command to let you clobber the CSS and appropriate SVG attributes with a custom colour scheme.

Requires https://github.com/facebook/react/pull/5781 to work, as otherwise react doesn't understand onLoad on object tags.

ara4n commented:
Could particularly do with review on Tinter.js - is it safe to store state in plain old vars like this, or should it be all bundled up into an exported object?  Or is that just hygiene?

richvdh commented on src/SlashCommands.js:
```diff
@@ -42,6 +43,11 @@ var commands = {
         return reject("Usage: /nick <display_name>");
     },
 
+    tint: function(room_id, args) {
```
(most of) the other commands have a comment saying what they do. wouldn't hurt here.

ara4n replied:
done

richvdh commented on src/components/structures/RoomView.js:
```diff
@@ -1193,8 +1197,8 @@ module.exports = React.createClass({
-                                <div className="mx_RoomView_tabCompleteEol">
-                                    <img src="img/eol.svg" width="22" height="16" alt="->|"/>
+                                <div className="mx_RoomView_tabCompleteEol" title="->|">
+                                    <object onLoad={ this.onSvgLoad } className="mx_Svg" type="image/svg+xml" data="img/eol.svg" width="22" height="16"/>
```
It would be great to factor out this whole <div> structure, complete with its onSvgLoad, to a separate react component so you don't have to c&p the idiom 150 times.

Kegsay replied:
+1

ara4n replied:
done

Kegsay commented:
>  is it safe to store state in plain old vars like this, or should it be all bundled up into an exported object? Or is that just hygiene?

It should be safe but in our case it isn't. It's usually safe because `require()` calls cache the first time they are called then return the *same instance* of `module.exports` again (along with closures containing those `vars`). It isn't safe for us because we tend to use the *same file* in *multiple projects* and we want that to refer to the same variables **cross-project**. This isn't how CommonJS modules work. You only get singleton-like behaviour in files in the same project. Cross-project references to the same file returns multiple instances (one for each project).

So what does this actually mean? It means that as it stands, if `vector-web` were to `require("matrix-react-sdk/lib/Tinter")` it would get an entirely new set of vars to the `Tinter` being required in `matrix-react-sdk`. You can refer to the same instance by doing a hack:

```javascript
module.exports = {
  ...
};

if (!global.mxTinter) { // this attaches to 'window' (!)
    global.mxTinter = module.exports;
}
module.exports = global.mxTinter;
```
You would need to move all your `var` declarations to be attached to `module.exports`. This is all 'orrible - and is filed as https://github.com/vector-im/vector-web/issues/464

richvdh commented on src/Tinter.js:
```diff
@@ -0,0 +1,214 @@
+            for (var k = 0; k < cssAttrs.length; k++) {
+                var attr = cssAttrs[k];
+                for (var l = 0; l < keyRgb.length; l++) {
+                    if (rule.style && rule.style[attr] === keyRgb[l]) {
```
can you add `if (!rule.style) continue;` at line 88, and simplify this?

ara4n replied:
yes, although it's very rare for a CSS rule to not have a style (it happens with weird rules like animation data which don't actually define styles), so it doesn't buy us much. have done it anyway.

Kegsay commented on src/Tinter.js:
```diff
@@ -0,0 +1,214 @@
+    }
+}
+
+function _onAction(payload) {
```
This file is *almost* entirely isolated from the rest of the project if it weren't for listening on the global dispatcher like this. I would prefer to do a direct function call to the `Tinter` from the new Tinter component you're writing as per:

```
<M-matthew> actually in fact if the svg thing was its own component it could probably handle it relatively elegantly
```

This matches other logic files which also have a set of public-facing methods.

Kegsay replied:
Also, as it stands, does this not race? Every tintable thing fires off a `svg_onload` event, and we don't seem to specify which things we want to tint anywhere so this will be tinting ALL THE THINGS! The only reason why it doesn't is because it doesn't get `svg_onload` events fired at it all the time?

Kegsay replied:
See https://github.com/matrix-org/matrix-react-sdk/pull/77#discussion_r48854434

ara4n replied:
Changed to call Tinter directly rather than via dispatch.  It doesn't tint 'all the things' as it passes the specific SVG DOM element to be tinted.

richvdh commented on src/Tinter.js:
```diff
@@ -0,0 +1,214 @@
+}
+
+var keyRgb = [
+    "rgb(118, 207, 166)",
```
magic numbers are magic

what do these correspond to?

Can we calculate these rather than hard-coding them?

ara4n replied:
Commented to clarify where they came from.  They can't really be calculated as they're constants describing shades of green.

richvdh commented on src/Tinter.js:
```diff
@@ -0,0 +1,214 @@
+// x = (255 - 234) / (255 - 118) = 0.16 
+
+var keyHex = [
+    "#76CFA6",
```
ditto

ara4n replied:
done

Kegsay commented on src/Tinter.js:
```diff
@@ -0,0 +1,214 @@
+    //     attr: name of the attribute to be clobbered, e.g. 'fill'
+    //     index: ordinal of primary, secondary
+    // }
+];
```
I think I finally grok how and why you're doing this. I don't like it because it is leaking DOM nodes all over the place because `Tinter` is a singleton when really the things it is manipulating are UI elements. Either strip this class to be pure-functional, or move it all to a separate UI component (the one you're writing). This means that you would have something like:

```javascript
React.createClass({

  _fixup: function(event) {
    // do fixup logic that Tinter.js is doing
  },

  render: function() {
    return <object onload={this._fixup} className="mx_Svg" type="image/svg+xml" data="img/leave.svg" />
});
```

This is much nicer since this will drop out of scope when the component is unmounted, stopping any leaks. My next questions would be:
 - Do you need `mx_Svg` anymore given you no longer need to do a full-scan of `document`?
 - You still need to retroactively apply tints to mounted svg components when people type `/tint`. This would be something that would be useful to be done via the `dispatcher` so you don't need to maintain a list of currently mounted svg components. So you'd have:
    ```javascript
    onAction: function(payload) {
      if (payload.action !== "tint_change") { return; }
      // apply new tint values from payload
    }
    ```

ara4n replied:
Agreed that this would be a nicer architecture and stop it from leaking DOM nodes everywhere.  However, how would the tint_change dispatch know where to find all of the tintable elements in order to tint them in the '// apply new tint values from payload' comment above?

I feel a bit inclined to move the proposed restructure here into a maintenance bug rather than invest more time in making the architecture perfect, given the leaks will be very minor.

ara4n commented:
I've addressed most of the feedback other than:
 * the big point about putting Tinter's state into modules.exports rather than plain old vars (given we're only ever going to be importing Tinter from matrix-react-sdk, I'm not sure it's worth the effort right now)
 * the big point about not leaking bits of SVG DOM into Tinter's state, but instead making Tinter functional and apply to the DOM components themselves.  Agreed this would be an improvement, but not sure it's worth the effort right now versus fixing other stuff.

Assuming we're agreed, I'll shift these to be maintenance bugs to address when we're out of crunch.

ara4n commented:
@kegsay PTAL

ara4n commented:
It's also worth noting that I was deliberately caching the required fixups for SVGs in the singleton Tinter rather than recalculating them every time onLoad (given fixups are still valid from previous loads).  Given SVGs could theoretically be quite big, wasting time re-walking them everytime they reload to generate fixups would chew CPU and time for no reason at all.

Kegsay commented:
> putting Tinter's state into modules.exports rather than plain old vars (given we're only ever going to be importing Tinter from matrix-react-sdk, I'm not sure it's worth the effort right now)

WFM. You've put a big warning up so I'm happy as is.

> not leaking bits of SVG DOM into Tinter's state, but instead making Tinter functional and apply to the DOM components themselves. Agreed this would be an improvement, but not sure it's worth the effort right now versus fixing other stuff.

So I had a look at how badly it leaks. My test was as follows:
 - Refresh page and swap to a room.
 - Take Heap Snapshot.
 - Change tint (if applicable)
 - Swap to different rooms ~10 times.
 - Take Heap Snapshot.

On `develop` - The heap snapshot delta was <1MB (21.6MB to 22.4MB).
On `matthew/dynamic-svg` - The heap snapshot delta was 34MB (24MB to 58MB).

This equates to about 3MB per room swap. This is too much for a long-lived web app. I don't mind fixing the leak for you if you're stretched for time, but I would be unhappy leaving it as-is.

ara4n commented:
i'm very surprised the leak is this bad. will fix it. suggestions on how to refer to the set of tintablesvgs when the user runs /tint?

> On 6 Jan 2016, at 12:01, Kegsay <notifications@github.com> wrote:
> 
> putting Tinter's state into modules.exports rather than plain old vars (given we're only ever going to be importing Tinter from matrix-react-sdk, I'm not sure it's worth the effort right now)
> 
> WFM. You've put a big warning up so I'm happy as is.
> 
> not leaking bits of SVG DOM into Tinter's state, but instead making Tinter functional and apply to the DOM components themselves. Agreed this would be an improvement, but not sure it's worth the effort right now versus fixing other stuff.
> 
> So I had a look at how badly it leaks. My test was as follows:
> 
> Refresh page and swap to a room.
> Take Heap Snapshot.
> Change tint (if applicable)
> Swap to different rooms ~10 times.
> Take Heap Snapshot.
> On develop - The heap snapshot delta was <1MB (21.6MB to 22.4MB).
> On matthew/dynamic-svg - The heap snapshot delta was 34MB (24MB to 58MB).
> 
> This equates to about 3MB per room swap. This is too much for a long-lived web app. I don't mind fixing the leak for you if you're stretched for time, but I would be unhappy leaving it as-is.
> 
> —
> Reply to this email directly or view it on GitHub.
>

Kegsay commented:
You just need to invert the control flow a bit. When people run /tint you should emit an event on the dispatcher (e.g. `tint_change`). Then you just need to listen for this in your `TintableSvg` and then change the tint. You never maintain the full set of tintable svgs.

Kegsay commented:
So the full flow would look like:

```javascript
// SlashCommands.js
tint: function(newColour) {
  Tinter.setTint(newColour);
  dispatcher.dispatch({
    action: "tint_change"
  });
}
// ============

// TintableSvg.js
componentWillMount: function() {
  dis.register(this._onAction);
},

componentDidMount: function() {
  Tinter.applyTint(ReactDOM.findDOMNode(this));
},

_onAction: function(payload) {
  if (payload.action !== "tint_change") { return; }
  Tinter.applyTint(ReactDOM.findDOMNode(this));
}

// ============

// Tinter.js
setTint: function(newColour) {
  currentColour = newColour;
},

applyTint: function(domNode) {
  // apply `currentColour` to `domNode`
}
```

And make sure that Tinter never persists the passed domNode. It should only persist the current colour.

ara4n commented:
oh, sorry, yes. was being thick.

ara4n commented:
0f52c0a fixes the leak by making TintableSvgs responsible for updating their own tints, and responding to tint_update dispatch events.  However I've kept the existing functions in Tinter.js rather than refactoring as per Kegan's full suggestion, as it retains symmetry with the half which is still needed for fixing up CSS (rather than SVGs), and i don't want to burn yet more time on this.

@kegsay: PTAL

Kegsay commented on src/components/views/elements/TintableSvg.js:
```diff
@@ -0,0 +1,70 @@
+
+    onAction: function(payload) {
+        if (payload.action !== 'tint_update') return;
+        Tinter.applySvgFixups(this.fixups);
```
If this is called prior to `onLoad` then this will NPE on `Tinter.js:192`.

ara4n replied:
why? this.fixups is initialised to [], not null?

Kegsay replied:
Oh, right you are. However, the way you're doing it isn't what you intend I think. You're defining `fixups` on `:27` as a *prototype* property, which is shared across all instances. You can't define *instance* properties like that, you need to use:

```javascript
componentWillMount: function() {
  this.fixups = [];
}
```

Without this, all `TintableSvg` elements would share the *same* `fixups` array(!)

ara4n replied:
oops, fixed, thanks

Kegsay commented:
LGTM - Did more heap dumps and now we no longer leak :)
